### PR TITLE
feat: 1422 autodetect new gcloud features

### LINKED ISSUE #1422: Autodetect new features in gcloud SDK
**Author the user story for this feature**

As a flank committer, I want to know when the new gcloud SDK feature is introduced so I can ensure the flank is fully compatible.

**Describe the solution**
An issue should be created when gcloud SDK introduces a new option (command, flag, etc) that is not implemented in the flank.

Possible solutions:
1. The easiest approach would be just to check if a new version of SDK was released and notify (via issue)
2. Manual checking may be time-consuming so ideally, we would have more detailed info. We can achieve this by tracking the release notes https://cloud.google.com/sdk/docs/release-notes#firebase_test_lab and create a new issue with info only related to test lab
3. The best would be to have very specific info on what was changed, missing etc. This approach needs to be investigating.

Once the solution is picked, provide detailed info in the documentation.

### PR DESCRIPTION
**Pull request by pawelpasterz**
Fixes #1422 

TLDR;
Fetch and parse `gcloud SDK` release notes to find new features/changes. If there are any -- create an epic with each new feature/change as a separate issue.

##### What it actually does?
1. Get the latest commit (current master sha)
1. Get previously checked commit (one that was previously 'latest')
1. Get new and old versions (fetch `VERSION` file) based on commits
1. If `new` == `old` -- no updates, exit process
1. Fetch `RELEASE_NOTES` file from the current master
1. Extract (with regex) section with chnages from `new` (included) till `old` (excluded)
1. Extract `Firebase Test Lab` changes only
1. Create a list of changes (each `*` is considered as a separate change/feature)
1. Create issues for each change
1. Create an issue with all changes and convert it to epic
1. Link 'change issues' to the created epic
1. If an epic (from the previous check) was not closed/finished -- do not create a new one, but update it instead

Example issue: https://github.com/pawelpasterz/flank/issues/107

Again, it's a bit tricky to test implementation. (As promised, it will be refactored)
You need to have:
* zenhub enabled for your repo (don't have to flank fork)
* zenhub repo id
* zenhub key
* PAT (with full access preferably)

Preparation:
* change zenhub repo id https://github.com/Flank/flank/blob/d1814271875824d7c4ae286566a0637a9e026ea6/flank-scripts/src/main/kotlin/flank/scripts/zenhub/ZenHubAPI.kt#L13
* change repo (your repo/fork) https://github.com/Flank/flank/blob/d1814271875824d7c4ae286566a0637a9e026ea6/flank-scripts/src/main/kotlin/flank/scripts/github/GithubApi.kt#L35
* create `gcloud SDK` label in your repo (might be not required, github can handle new labels somehow(?) )
* enable creating issues in your repo

1. `./gradlew assemble`
1.  run `java -jar ./flank-scripts/build/libs/flank-scripts.jar shell firebase checkForSdkUpdate --github-token=[github token] --zenhub-token=[zenhub token]`
1. Depends of what repo (fork or not) did you use, there might be info about workflow found or not. But it should end with `No new commits`
    ```
    ** Find previously checked commit
    ** No workflow run found for update_dependencies_and_client.yml
    ** Commit found:
          SHA: 8de5b70bc8996687e097d604815018c7b60f29d3
          timestamp: 2020-12-16T04:09:10Z
    ** Find latest commit
    ** Commit found:
          SHA: 8de5b70bc8996687e097d604815018c7b60f29d3
          timestamp: 2020-12-16T04:09:10Z
    ** No new commits since the last run
    ```
1. replace https://github.com/Flank/flank/pull/1436/files#diff-45b4dc40914b7eba78ddf9e571c0f217266d4fb7e544552d2bfcaa2c7a94fb1dR23 with `2020-09-16T04:09:10Z`
1. re-assemble and run, the output should look like one below + issues are created (of course, links and numbers will differ)
    ```
    ** Find previously checked commit
    ** Commit found:
          SHA: 43cc3f11d80e138f456a002f0481c65aea092902
          timestamp: 2020-09-16T04:07:35Z
    ** Find latest commit
    ** Commit found:
          SHA: 8de5b70bc8996687e097d604815018c7b60f29d3
          timestamp: 2020-12-16T04:09:10Z
    ** No opened issue
    ** New version 321.0.0
    ** Old version 310.0.0
    ** Create linked issues
    ** Issue created:
         url:    https://github.com/pawelpasterz/flank/issues/98
         number: 98
    ** Issue created:
         url:    https://github.com/pawelpasterz/flank/issues/99
         number: 99
    ** Issue created:
          url:    https://github.com/pawelpasterz/flank/issues/100
          number: 100
    ** Issue created:
         url:    https://github.com/pawelpasterz/flank/issues/101
         number: 101
    ** Create new epic
    ** Issue created:
         url:    https://github.com/pawelpasterz/flank/issues/102
         number: 102
    ** Issue 102 successfully converted to an epic
    ```
1.  change the date again but this time with `2020-10-16T04:09:10Z`
1. change https://github.com/Flank/flank/pull/1436/files#diff-1a833b8e4991c1e9615c5634567625ff14a5ca97672158fe6352f37b29b6794cR10 with your username
1. re-assemble and run
    ```
    ** Find previously checked commit
    ** Commit found:
          SHA: e58409d11d3079338a2f01ef70d8e1fb283b3977
          timestamp: 2020-09-30T04:08:19Z
    ** Find latest commit
    ** Commit found:
          SHA: 8de5b70bc8996687e097d604815018c7b60f29d3
          timestamp: 2020-12-16T04:09:10Z
    ** Issue found: https://github.com/pawelpasterz/flank/issues/107
    ** New version 321.0.0
    ** Old version 312.0.0
    ** Create linked issues
    ** Issue created:
         url:    https://github.com/pawelpasterz/flank/issues/108
         number: 108
    ** Issue created:
         url:    https://github.com/pawelpasterz/flank/issues/109
         number: 109
    ** Issue created:
         url:    https://github.com/pawelpasterz/flank/issues/110
         number: 110
    ** Issue created:
         url:    https://github.com/pawelpasterz/flank/issues/111
         number: 111
    ** Update existing epic
    ** Issues successfully added to the issue 107
    ```

## Checklist

- [x] Documented
- [x] Unit tested

### CODE REVIEW

File: flank-scripts/build.gradle.kts [hunk 1]
```diff
@@ -28,7 +28,7 @@ shadowJar.apply {
     }
 }
 // <breaking change>.<feature added>.<fix/minor change>
-version = "1.2.9"
+version = "1.3.0"
```
Comment (Sloox) [inline flank-scripts/build.gradle.kts:31]: Will have a conflict here with https://github.com/Flank/flank/pull/1428

File: flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CheckForSDKUpdateCommand.kt [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    help = "Verifies if there were changes in gcloud sdk that need to be implemented in flank"
+) {
+
+    private val githubToken by option(help = "Git Token").required()
+    private val zenhubToken by option(help = "Zenhub Token").required()
+    private val lastRun by lazy { runBlocking { getLastSDKUpdateRunDate(githubToken) } }
+    private val openedUpdates by lazy { runBlocking { checkForOpenedUpdates(githubToken) } }
+
+    override fun run() {
+        runBlocking {
```
Comment (Sloox) [inline flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CheckForSDKUpdateCommand.kt:27]:
```suggestion
     override fun run() =  runBlocking {...
```
?

File: flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/OpenedUpdates.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import flank.scripts.github.getGitHubIssueList
+
+suspend fun checkForOpenedUpdates(token: String) = getGitHubIssueList(
+    githubToken = token,
+    parameters = listOf(
+        "creator" to "github-actions[bot]",
+        "state" to "open",
+        "labels" to "gcloud SDK"
+    )
+)
```
Comment (Sloox) [inline flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/OpenedUpdates.kt:14]: odd spacing here
Feels like there is one new line too many

File: flank-scripts/src/main/kotlin/flank/scripts/zenhub/ZenHubAPI.kt [hunk 4]
```diff
@@ -8,11 +8,14 @@ import com.github.kittinunf.result.getOrNull
...
 import com.github.kittinunf.result.success
 import flank.scripts.utils.toJson
+import flank.scripts.zenhub.objects.ConvertToEpicRequest
+import flank.scripts.zenhub.objects.UpdateEpicRequest
 import kotlinx.serialization.Serializable
 
 const val FLANK_REPO_ID = 84221974
 internal const val ZENHUB_BASE_URL = "https://api.zenhub.com/p1/repositories/$FLANK_REPO_ID"
 
+// GET
```
Comment (Sloox) [inline flank-scripts/src/main/kotlin/flank/scripts/zenhub/ZenHubAPI.kt:18]: Why not change the function name instead?

File: flank-scripts/src/main/kotlin/flank/scripts/zenhub/ZenHubAPI.kt [hunk 5]
```diff
@@ -26,6 +29,7 @@ private suspend fun getEstimation(zenhubToken: String, issueNumber: Int) =
         .onError { println("Could not get estimations because of ${it.message}") }
         .getOrNull()
 
+// PUT
```
Comment (Sloox) [inline flank-scripts/src/main/kotlin/flank/scripts/zenhub/ZenHubAPI.kt:32]: Same here

**Review by Sloox — CHANGES_REQUESTED** [review]

File: flank-scripts/build.gradle.kts (hunk 1 above)
Comment (pawelpasterz) [reply flank-scripts/build.gradle.kts:31]: True

File: flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/OpenedUpdates.kt (hunk 3 above)
Comment (pawelpasterz) [reply flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/OpenedUpdates.kt:14]: `ktlint` formatted that way.

File: flank-scripts/src/main/kotlin/flank/scripts/zenhub/ZenHubAPI.kt (hunk 4 above)
Comment (pawelpasterz) [reply flank-scripts/src/main/kotlin/flank/scripts/zenhub/ZenHubAPI.kt:18]: Some method names clearly explain their purpose. Adding HTTP prefix name would look strange (in the end it is not HTTP client library) example: `convertIssueToEpic` -> `postConvertIssueToEpic`

But I think you are right and those `// [HTTP method]` comments are redundant. I'll remove them in `GitHubApi.kt` as well

**Review by dstiers11 — APPROVED** [review]

File: flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CheckForSDKUpdateCommand.kt (hunk 2 above)
Comment (pawelpasterz) [reply flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CheckForSDKUpdateCommand.kt:27]: Sure, I can change that

File: .github/workflows/update_dependencies_and_client.yml [hunk 6]
```diff
@@ -94,3 +94,18 @@ jobs:
             firease_api
           reviewers: bootstraponline,jan-gogo,pawelpasterz,adamfilipow92,piotradamczyk5,Sloox,axelzuziak-gogo
           draft: false
+
+  check_for_sdk_updates:
+    runs-on: macos-latest
+    needs: [ get_token_and_date ]
```
Comment (piotradamczyk5) [inline .github/workflows/update_dependencies_and_client.yml:100]: You are using `secrets.GITHUB_TOKEN ` so I do not think that you need to wait for `get_token_and_date` job

File: flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CheckForSDKUpdateCommand.kt [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                else -> run {
+                    println("** No new features")
+                    return@runBlocking
+                }
+            }
+        }
+    }
+
+    private fun createContext(sha: String) = SDKUpdateContext(
+        oldVersion = parseToVersion(Fuel.get("$RAW_GITHUB/$sha/google-cloud-sdk/VERSION").responseString().third.get()),
```
Comment (piotradamczyk5) [inline flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CheckForSDKUpdateCommand.kt:58]: I will wrap it to function

File: flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CommitList.kt [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    .maxByOrNull { Instant.from(DateTimeFormatter.ISO_INSTANT.parse(it.commit.author.date)) }
+    .also {
+        if (it != null)
+            println(
+                """
+                |** Commit found: 
+                |      SHA: ${it.sha}
+                |      timestamp: ${it.commit.author.date}""".trimMargin()
+            )
+    }?.sha
```
Comment (piotradamczyk5) [inline flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CommitList.kt:26]: could you make similiar as before
`    .also(::logCommitFound)`

File: flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CheckForSDKUpdateCommand.kt [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            println("** New version $newVersion")
+            println("** Old version $oldVersion")
+            when {
+                oldVersion < newVersion && openedUpdates != null -> updateOpenedEpic()
+                oldVersion < newVersion && openedUpdates == null -> createEpicIssue()
+                else -> run {
+                    println("** No new features")
+                    return@runBlocking
+                }
+            }
```
Comment (piotradamczyk5) [inline flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CheckForSDKUpdateCommand.kt:53]: We could move `oldVersion < newVersion` to top to make code smaller

```kotlin
when {
    oldVersion >= newVersion -> run {
                    println("** No new features")
                    return@runBlocking
                }
    openedUpdates != null -> updateOpenedEpic()
    openedUpdates == null -> createEpicIssue()
            }
`

File: flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/Extensions.kt [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    val (updates, subIssues) = createSubIssues()
+    requireNotNull(openedIssue)
+
+    println("** Update existing epic")
+    patchIssue(
+        githubToken = githubToken,
+        issueNumber = openedIssue.number,
+        payload = GitHubUpdateIssueRequest(
+            title = "Implement new gcloud features [$newVersion]",
+            body = openedIssue.body + "\n" + updates
```
Comment (piotradamczyk5) [inline flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/Extensions.kt:45]: move it to const val and put it on `:common` module please

File: flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CheckForSDKUpdateCommand.kt (hunk 9 above)
Comment (pawelpasterz) [reply flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CheckForSDKUpdateCommand.kt:53]: Good idea, forgot `when` is not an expression in that case and don't have to be exhaustive 😅

File: flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CheckForSDKUpdateCommand.kt (hunk 7 above)
Comment (pawelpasterz) [reply flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CheckForSDKUpdateCommand.kt:58]: Sure, thanks!

File: .github/workflows/update_dependencies_and_client.yml (hunk 6 above)
Comment (pawelpasterz) [reply .github/workflows/update_dependencies_and_client.yml:100]: Good point. 
BTW, do you think I could remove `create gh token` at all? It's used 
https://github.com/Flank/flank/blob/d1814271875824d7c4ae286566a0637a9e026ea6/.github/workflows/update_dependencies_and_client.yml#L49
and here 
https://github.com/Flank/flank/blob/d1814271875824d7c4ae286566a0637a9e026ea6/.github/workflows/update_dependencies_and_client.yml#L86
and can be replaced with `secrets.GITHUB_TOKEN` ?

File: flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CommitList.kt (hunk 8 above)
Comment (pawelpasterz) [reply flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/CommitList.kt:26]: Sure, why not.

File: flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/Extensions.kt (hunk 10 above)
Comment (pawelpasterz) [reply flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/Extensions.kt:45]: Hm...I think I don't fully understand. Both `openedIssue.body` and `updates` are variables.

File: .github/workflows/update_dependencies_and_client.yml (hunk 6 above)
Comment (piotradamczyk5) [reply .github/workflows/update_dependencies_and_client.yml]: no you could not use `secrets.GITHUB_TOKEN` in those cases.
The reason why we are using a custom token (from Github App) is that we create PR with changes (default Github token can not trigger workflows)

File: flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/Extensions.kt (hunk 10 above)
Comment (piotradamczyk5) [reply flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/Extensions.kt:45]: I mean new line `\n`😄

**Review by piotradamczyk5 — APPROVED** [review]

File: flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/Extensions.kt (hunk 10 above)
Comment (pawelpasterz) [reply flank-scripts/src/main/kotlin/flank/scripts/shell/firebase/sdk/Extensions.kt:45]: 🤦  arghh

File: .github/workflows/update_dependencies_and_client.yml (hunk 6 above)
Comment (pawelpasterz) [reply .github/workflows/update_dependencies_and_client.yml]: I see, thanks for explanation

**Review by Sloox — APPROVED** [review]
👍
